Add support for Yul ASTs output (#14177)
ethereum/solidity: Changelog.md

Compiler Features:

>  * Commandline Interface: Add ``--ast-compact-json`` option with Yul input (e.g. with ``--strict-assembly``, ``--yul``)

Reviewer: ```suggestion
 * Commandline Interface: Add ``--ast-compact-json`` option with Yul input (e.g. with ``--strict-assembly``, ``--yul``).
```

>  * Commandline Interface: Add ``--ir-ast-json`` and ``--ir-optimized-ast-json`` output options which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.

Reviewer 2: I think that the removal of the suffix was meant to apply to CLI options too. The `-json` here is just as superfluous.

```suggestion
 * Commandline Interface: Add ``--ir-ast`` and ``--ir-optimized-ast`` output options which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.
```

Reviewer 3: It's consistent with other options, so let's keep it as is.

>  * Standard JSON Interface: Support "irAstJson" amd "irOptimizedAstJson" options in output selection which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.

Reviewer: ```suggestion
 * Standard JSON Interface: Support ``irAstJson`` amd ``irOptimizedAstJson`` options in output selection which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.
```

Reviewer 4: Thanks for the changes @GiokaMarkella, in our regular meeting we decided that the `irAstJson` and `irOptimizedAstJson` should be renamed to `irAst` and `irOptimizedAst` respectively. I.e., just remove the `Json` part.

Could you please made the change? Then the PR should be ready to be merged :)

>  * Standard JSON Interface: Support "irAstJson" amd "irOptimizedAstJson" options in output selection which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.

Reviewer 4: ```suggestion
 * Standard JSON Interface: Support "irAstJson" and "irOptimizedAstJson" options in output selection which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.
```

>  * Standard JSON Interface: Support the "ast" option in output selection with Yul input.

Reviewer: ```suggestion
 * Standard JSON Interface: Support the ``ast`` option in output selection with Yul input.
```

>  * Commandline Interface: Add ``--ast-compact-json`` option with Yul input (e.g. with ``--strict-assembly``, ``--yul``).
>  * Commandline Interface: Add ``--ir-ast-json`` and ``--ir-optimized-ast-json`` output options which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.
>  * Standard JSON Interface: Support ``irAst`` and ``irOptimizedAst`` options in output selection which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.
>  * Standard JSON Interface: Support the ``ast`` option in output selection with Yul input.
> 

Reviewer 2: I'm going nitpick these a bit too :)

We should make it clearer what is completely new and what just supported in extra input modes.

```suggestion
 * Commandline Interface: Add ``--ir-ast-json`` and ``--ir-optimized-ast-json`` outputs, providing AST in compact JSON format for IR and optimized IR.
 * Commandline Interface: Support ``--ast-compact-json`` output also in assembler mode.
 * Standard JSON Interface: Add ``ast`` contract-level output for Yul input.
 * Standard JSON Interface: Add ``irAst`` and ``irOptimizedAst`` contract-level outputs for Solidity input, providing AST in compact JSON format for IR and optimized IR.
```

 * SMTChecker: Properties that are proved safe are now reported explicitly at the end of the analysis. By default, only the number of safe properties is shown. The CLI option ``--model-checker-show-proved-safe`` and the JSON option ``settings.modelChecker.showProvedSafe`` can be enabled to show the full list of safe properties.
 * SMTChecker: Group all messages about unsupported language features in a single warning. The CLI option ``--model-checker-show-unsupported`` and the JSON option ``settings.modelChecker.showUnsupported`` can be enabled to show the full list.
 * Yul EVM Code Transform: If available, use ``push0`` instead of ``codesize`` to produce an arbitrary value on stack in order to create equal stack heights between branches.

>  * Commandline Interface: Add ``--ast-compact-json`` option with Yul input (e.g. with ``--strict-assembly``, ``--yul``)

Reviewer: Just a note: you're gonna have to rebase and move these entries up to the `0.8.21` section, since we released a couple of days ago. Also, the entries should be alphabetically sorted.

>  * Commandline Interface: Add ``--ir-ast-json`` and ``--ir-optimized-ast-json`` output options which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.

Reviewer 4: This option seems to be enabled by default when using `--strict-assembly` or `--yul` (e.g. `solc --yul t.yul`). It should only output the `Yul-IR` AST when requested, no?

 * SMTChecker: Properties that are proved safe are now reported explicitly at the end of analysis. By default, only the number of safe properties is shown. The CLI option ``--model-checker-show-proved-safe`` and the JSON option ``settings.modelChecker.showProvedSafe`` can be enabled to show the full list of safe properties.
 * Standard JSON Interface: Add experimental support for importing ASTs via Standard JSON.
 * Yul EVM Code Transform: If available, use ``push0`` instead of ``codesize`` to produce an arbitrary value on stack in order to create equal stack heights between branches.

>  * Commandline Interface: Add ``--ast-compact-json`` option with Yul input (e.g. with ``--strict-assembly``, ``--yul``)
>  * Commandline Interface: Add ``--ir-ast-json`` and ``--ir-optimized-ast-json`` output options which output the Yul-IR and the optimized Yul-IR in AST compact JSON format.

Reviewer: ```suggestion
```

Author: Could you explain the reason for this suggestion?

Reviewer: This is the `0.8.20` section if I'm not mistaken?

Reviewer 3: Haha, I didn't understand that comment at all myself either :-). But yeah, after a rebase the changelog entries should move to the new 0.8.21 section now that we have 0.8.20 released.

Author: Oh sorry! Thought I had deleted that haha

 * Standard JSON Interface: Add experimental support for importing ASTs via Standard JSON.
 * Yul EVM Code Transform: If available, use ``push0`` instead of ``codesize`` to produce an arbitrary value on stack in order to create equal stack heights between branches.

> 

Reviewer 2: The spacing here was fine. Please don't change it, it's just noise - someone will just add it again in another PR.
